machina/loss_functional.py

```diff
@@ -296,6 +296,151 @@ def sac(pol, qfs, targ_qfs, log_alpha, batch, gamma, sampling=1, reparam=True, n
     return pol_loss, qf_losses, alpha_loss
 
 
+def r2d2_sac(pol, qfs, targ_qfs, log_alpha, batch, gamma, sampling=1, burn_in_length=40, reparam=True, normalize=False, eps=1e-6):
+    """
+    Loss for soft actor critic.
+
+    Parameters
+    ----------
+    pol : Pol
+    qfs : list of SAVfunction
+    targ_qfs : list of SAVfunction
+    log_alpha : torch.Tensor
+    batch : dict of torch.Tensor
+    gamma : float
+    sampling : int
+        Number of samping in calculating expectation.
+    burn_in_length : int
+        Length of batches for burn-in.
+    reparam : bool
+        Reparameterization trick is used or not.
+    normalize : bool
+        If True, normalize value of log likelihood.
+    eps : float
+
+    Returns
+    -------
+    pol_loss, qf_loss, alpha_loss, td_losses : torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor
+    """
+
+    # trajectories for burn-in
+    bi_obs = batch['obs'][:burn_in_length]
+    bi_next_obs = batch['next_obs'][:burn_in_length]
+    bi_acs = batch['acs'][:burn_in_length]
+    bi_h_masks = batch['h_masks'][:burn_in_length]
+    bi_next_h_masks = batch['h_masks'][1:burn_in_length+1]
+
+    # trajectories for train
+    obs = batch['obs'][burn_in_length: -1]
+    acs = batch['acs'][burn_in_length: -1]
+    rews = batch['rews'][burn_in_length: -1]
+    next_obs = batch['next_obs'][burn_in_length: -1]
+    dones = batch['dones'][burn_in_length: -1]
+    h_masks = batch['h_masks'][burn_in_length: -1]
+    next_h_masks = batch['h_masks'][burn_in_length+1:]
+
+    # hidden states (time_seq, batch_size, *, cell_size)
+    a_hs = (batch['hs'][0, :, 0], batch['hs'][0, :, 1])
+    next_a_hs = batch['hs'][1]
+    q_hs = [(batch['q_hs'+str(i)][0, :, 0], batch['q_hs'+str(i)][0, :, 1])
+            for i in range(len(qfs))]
+    targ_q_hs = [(batch['targ_q_hs'+str(i)][0, :, 0],
+                  batch['targ_q_hs'+str(i)][0, :, 1]) for i in range(len(targ_qfs))]
+
+    alpha = torch.exp(log_alpha)
+    train_length = batch['obs'].size()[0] - burn_in_length - 1
+    time_seq, batch_size, *_ = batch['obs'].size()
+
+    pol.hs = a_hs
+    with torch.no_grad():
+        bi_pd_params = [pol(bi_obs[i:i+1], h_masks=bi_h_masks[i:i+1])[-1]
+                        for i in range(burn_in_length)]
```

Comment:
Here is code example
```
_bi_pd_params = pol(bi_obs, h_masks=bi_h_masks)[-1]
separated_bi_pd_params = [_bi_pd_params[key] for key in sorted(_bi_pd_params.keys())]
bi_pd_params = []
for params in zip(*separated_bi_pd_params):
    params_dict = {key: param for key, param in zip(sorted(_bi_pd_params.keys()), params)}
    bi_pd_params.append(params_dict)
```

I think this may be faster...